source/Nevermore/Advanced/WriteTransaction.cs

```diff
@@ -217,5 +236,60 @@ public async Task CommitAsync(CancellationToken cancellationToken = default)
             await Transaction.CommitAsync(cancellationToken);
             await configuration.Hooks.AfterCommitAsync(this);
         }
+
+        object[] ExecuteDataModification(PreparedCommand command)
+        {
+            if (!command.Mapping.IsRowVersioningEnabled)
+            {
+                ExecuteNonQuery(command);
+                return Array.Empty<object>();
+            }
+
+            //The results need to be read eagerly so errors are raised while code is still executing within CommandExecutor error handling logic
+            return ReadResults(command, reader => reader.GetValue(0));
+        }
+
+        object ExecuteSingleDataModification(PreparedCommand command)
+        {
+            var results = ExecuteDataModification(command);
+            return results.SingleOrDefault();
+        }
+
+        async Task<object[]> ExecuteDataModificationAsync(PreparedCommand command, CancellationToken cancellationToken)
+        {
+            if (!command.Mapping.IsRowVersioningEnabled)
+            {
+                await ExecuteNonQueryAsync(command, cancellationToken);
+                return Array.Empty<object>();
+            }
+
+            return await ReadResultsAsync(command, reader => reader.GetValue(0));
+        }
+
+        async Task<object> ExecuteSingleDataModificationAsync(PreparedCommand command, CancellationToken cancellationToken)
+        {
+            var results = await ExecuteDataModificationAsync(command, cancellationToken);
+            return results.SingleOrDefault();
+        }
+
+        void ApplyNewRowVersionsIfRequired(IReadOnlyList<object> documentList, DocumentMap mapping, object[] newRowVersions)
+        {
+            if (!mapping.IsRowVersioningEnabled) return;
+
+            for (var i = 0; i < documentList.Count; i++)
+            {
+                ApplyNewRowVersionIfRequired(documentList[i], mapping, newRowVersions[i]);
+            }
+        }
+
+        void ApplyNewRowVersionIfRequired<TDocument>(TDocument document, DocumentMap mapping, object newRowVersion) where TDocument : class
+        {
+            if (!mapping.IsRowVersioningEnabled) return;
+
+            if (newRowVersion == null) throw new StaleDataException($"Modification failed for '{typeof(TDocument).Name}' document with '{mapping.GetId(document)}' Id because submitted data was out of date. Refresh the document and try again.");
```

Comment:
hmm ... reusing these two private methods here in both `INSERT` and `UPDATE` operations somehow makes me feel it's less readable since by just reading through the code line by line, it shows that `INSERT` operation could throw this exception as well, however, in theory, this exception will only be thrown for `UPDATE`.

Reply:
INSERT can throw here if the new row version is not returned, e.g. a bug in the SQL generation